### ISSUE TITLE
How to SetDirty on all instances of an Azure function app?

### ISSUE THREAD

**Issue by ms-amakumar** [Azure/AppConfiguration#549]
I have an Azure function app deployed on 3 instances of App service plan, and I am using a httptrigger to SetDirty on Configuration Refresher. But as it turns out, configuration is only being refreshed on the original instance where SetDirty was executed. Other two instances are still returning the old config value. Even if I use event grid for app config events, wouldn't that also send the notification just to one of my instances and not all. How do you suggest we should handle this?

       public class ShowMessage
       {

        private readonly IConfiguration _configuration;
        private readonly IConfigurationRefresher _configurationRefresher;

        public ShowMessage(IConfiguration configuration, IConfigurationRefresherProvider refresherProvider)
        {
            _configuration = configuration;
            _configurationRefresher = refresherProvider.Refreshers.First();
        }

        [FunctionName("ShowMessage")]
        public async Task<IActionResult> Run(
            [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post", Route = null)] HttpRequest req,
            ExecutionContext exCtx,
            ILogger log)
        {
            string instanceId = Environment.GetEnvironmentVariable(
            "WEBSITE_INSTANCE_ID",
            EnvironmentVariableTarget.Process);

            // Signal to refresh the configuration if the registered key(s) is modified.
            // This will be a no-op if the cache expiration time window is not reached.
            // The configuration is refreshed asynchronously without blocking the execution of the current function.
            _ = _configurationRefresher.TryRefreshAsync();

            string key = "TestApp:Settings:Message";
            string message = $"{ _configuration[key]}: {instanceId}: {DateTime.UtcNow.ToLongTimeString()}";

            log.LogInformation(message);
            return new OkObjectResult(message);
        }

        [FunctionName("updateconfig")]
        public async Task<IActionResult> RunConfigUpdate(
            [HttpTrigger(AuthorizationLevel.Function, "get", "post", Route = null)] HttpRequest req,
            ILogger log)
        {
            string instanceId = Environment.GetEnvironmentVariable(
            "WEBSITE_INSTANCE_ID",
            EnvironmentVariableTarget.Process);
            _configurationRefresher.SetDirty(TimeSpan.FromSeconds(1));
            await Task.Delay(3000);
            _ = _configurationRefresher.TryRefreshAsync();
            return new OkObjectResult($"Done: {instanceId}: {DateTime.UtcNow.ToLongTimeString()}");
        }
    }

    class Startup : FunctionsStartup
    {
        public override void ConfigureAppConfiguration(IFunctionsConfigurationBuilder builder)
        {
            // Add Azure App Configuration as additional configuration source
            builder.ConfigurationBuilder.AddAzureAppConfiguration(options =>
            {
                options.Connect("*****")
                       // Load all keys that start with `TestApp:`
                       .Select("TestApp:*")
                       // Configure to reload configuration if the registered 'Sentinel' key is modified
                       .ConfigureRefresh(refreshOptions =>
                            refreshOptions.Register("TestApp:Settings:Sentinel", refreshAll: true).SetCacheExpiration(TimeSpan.FromDays(1))
                        )
                       // Indicate to load feature flags
                       .UseFeatureFlags(_ => _.CacheExpirationInterval = TimeSpan.FromDays(1));
            });
        }

        public override void Configure(IFunctionsHostBuilder builder)
        {
            // Make Azure App Configuration services and feature manager available through dependency injection
            builder.Services.AddAzureAppConfiguration();
            builder.Services.AddFeatureManagement();
        }
    }

**Comment by zhenlan** [comment]
@ms-amakumar when you have multiple instances of your application (Azure Functions in this case), you will need a way so the push notification events can be sent to all instances. Here is an example that uses Service Bus for this purpose.
https://docs.microsoft.com/en-us/azure/azure-app-configuration/enable-dynamic-configuration-dotnet-core-push-refresh

It does take an extra complexity to implement the push model properly, but it's worthwhile when you have _many_ instances. When your instance number is small, it is simpler to implement the poll model.

**Comment by ms-amakumar** [comment]
@zhenlan 

Still running into the same issue. Here is what I tried.
Implemented service bus handler and have 3 instances of the func app running. Updated a key, followed by Sentinel key. These two update notifications only went to 2 of the 3 instances and marked configurationRefresher dirty, but the other two instances are still returning the old value as the service bus notification never reached to these instances.

        [FunctionName("AppConfigNotifications")]
        public void RunNotifications([ServiceBusTrigger("testtopic", "mysubscription", Connection = "ServiceBusConnectionString")] string messageText, ILogger log)
        {
            string instanceId = Environment.GetEnvironmentVariable(
            "WEBSITE_INSTANCE_ID",
            EnvironmentVariableTarget.Process);

            //string messageText = Encoding.UTF8.GetString(message.Body);
            JsonElement messageData = JsonDocument.Parse(messageText).RootElement.GetProperty("data");
            string key = messageData.GetProperty("key").GetString();
            dirtyMessage.Add($"Event received for Key = {key}: {instanceId}");
            Console.WriteLine(string.Join(",", dirtyMessage));
            _configurationRefresher.SetDirty();
            log.LogInformation($"C# ServiceBus topic trigger function processed message: {instanceId}:{string.Join(",", dirtyMessage)}");
        }

**Comment by zhenlan** [comment]
@ms-amakumar I think you will need to create a separate service bus subscription for each Function instance.

**Comment by ms-amakumar** [comment]
@zhenlan  Is there a documentation I can follow? As those function instances are created dynamically it'll be tricky to create on demand subscription. Haven't seen that being done before.

**Comment by RichardMuniu** [comment]
Hi @ms-amakumar! 
Here’s the link to the [Service Bus Management Client reference](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.servicebus.management.managementclient?view=azure-dotnet). Additionally, here's a code sample of how you can create and use dynamic subscriptions.

```
private static async Task<SubscriptionDescription> CreateDynamicSubscription()
{
    // initialize a management client
    _managementClient = new ManagementClient(_serviceBusConnectionString);
    // create dynamic subscription name, for instance based on the machine name 
    string subscriptionName = $"{System.Environment.MachineName}-dynamic";

    . . .

    // check if subscription already exists
    var subscriptionExists = await _managementClient.SubscriptionExistsAsync(_serviceBusTopic, subscriptionName);
    if (subscriptionExists)
    {
        _logger.LogInformation($"Subscription '{subscriptionName}' already exists");
        return await _managementClient.GetSubscriptionAsync(_serviceBusTopic, subscriptionName);
    }
    _logger.LogInformation($"Creating subscription '{subscriptionName}'");
    return await _managementClient.CreateSubscriptionAsync(_serviceBusTopic, $"{subscriptionName}");
}
````

If you'd like to programmatically clean up your created subscriptions at some point after creation, you can use the `DeleteSubscriptionAsync` method for that.

`managementClient.DeleteSubscriptionAsync(_serviceBusTopic, _subscriptionName);`

Let me know if you have any questions!

**Comment by zhenlan** [comment]
@ms-amakumar I hope @RichardMuniu answered your question. Closing.

**Comment by inishantmishra** [comment]
@zhenlan @RichardMuniu 

Are you guys are saying, that if any feature flag is modified then to read the events from event grid and to update IConfiguration for multiple instances of azure function, do we need to create multiple service bus queues and subscriptions????

Can you please explain in detail how to do and what if we are creating instances dynamically as per the dynamic scaling of azure functions? Does ProcessPushNotification not update IConfiguration for all instances?
This is a very basic requirement to update all instances for feature flags.

**Comment by zhenlan** [comment]
@inishantmishra it involves non-trivial work to implement the push model correctly and it can be costly if you don't use required dependent services (eg. service bus or event hub) already. At this point, I'd recommend using the poll model for all scenarios. Both App Configuration service and its client libraries (configuration providers) have been optimized for the poll model. It's much simpler for customers to implement and it won't add extra complexity to customers' service architecture.

To implement the poll model in Azure Functions, please see the following document:
https://learn.microsoft.com/en-us/azure/azure-app-configuration/enable-dynamic-configuration-azure-functions-csharp?tabs=isolated-process

**Comment by inishantmishra** [comment]
@zhenlan @drago-draganov 
I tried with Poll method too but its not working what documentation has mentioned.
What I did.
1> I have 2 feature flags defined- FeatureFlagA & FeatureFlagB both are disabled.
2> I have a Sentinel key defined with name- sentinel which is disabled too.

Now, In my azure function which is in .net 8 dotnet-isolated, I have defined as following

`builder.AddAzureAppConfiguration(options =>
{
    options.Connect(configuration.GetSection("ConnectionStrings:AppConfig").Value)
        .Select(KeyFilter.Any, LabelFilter.Null)
        .ConfigureRefresh(refresh =>
        {
            refresh.Register("sentinel", refreshAll: true);
        })
        .UseFeatureFlags();
});`

Now, when Azurefuction gets started at first, it loads all Fetaure Flags values.. Hence it will load
FeatureFlagA as false
FeatureFlagB as false

Since, we haven't defined any cache interval, then it will take 30 seconds as default to check sentinel key.

Now, lets say I went ahead and enabled the FeatureFlagA.
Post, 30seconds, What I observed was that, even if i haven't modified or touched the sentinel key, feature flags values were updated with latest values.
I got following values-
FeatureFlagA as true
FeatureFlagB as false

So, it means, after 30 seconds it is refreshing all feature flags and not even monitoring/checkign sentinel key. Hence, it seems there is no usage of sentinel key.


I also did another way where I tried adding the cache expiration to 2 hours for FeatureFlags and left the sentinelKey Expiration as 30 seconds default only, like below-
`
builder.AddAzureAppConfiguration(options =>
{
    options.Connect(configuration.GetSection("ConnectionStrings:AppConfig").Value)
        .Select(KeyFilter.Any, LabelFilter.Null)
        .ConfigureRefresh(refresh =>
        {
            refresh.Register("sentinel", refreshAll: true);
        })
        .UseFeatureFlags(f =>
        {
            f.Select(KeyFilter.Any, LabelFilter.Null);
            f.CacheExpirationInterval = TimeSpan.FromHours(2);
        });
});
`
What I observed with above was that, values were getting updated only after 2 hours. Nothing was getting changed/updated after the sentinel key change post 30 seconds.


So By doing both the way, I found that there is no usage of sentinelKey. It only respects the FeatureFlag CacheExpiration time and it will refreshes all feature flags with latest values as per this cache.


If we remove the UseFeatureFlags, then Azure function wont even loads the feature flags values on start and wont even change post modification (including sentinel key modification too).


Can anyone help me, how to make sure feature flags wont get updated automatically but only when sentinel key is changed.
